src/core/pointcloud/qgspointclouddataprovider.cpp

```diff
@@ -176,3 +183,145 @@ QVariant QgsPointCloudDataProvider::metadataClassStatistic( const QString &, con
 {
   return QVariant();
 }
+
+static void _pointXY( QgsPointCloudRenderContext &context, const char *ptr, int i, double &x, double &y )
+{
+  const qint32 ix = *reinterpret_cast< const qint32 * >( ptr + i * context.pointRecordSize() + context.xOffset() );
+  const qint32 iy = *reinterpret_cast< const qint32 * >( ptr + i * context.pointRecordSize() + context.yOffset() );
+  x = context.offset().x() + context.scale().x() * ix;
+  y = context.offset().y() + context.scale().y() * iy;
+}
+
+/**
+     * Retrieves the z value for the point at index \a i.
+     */
+static double _pointZ( QgsPointCloudRenderContext &context, const char *ptr, int i )
+{
+  const qint32 iz = *reinterpret_cast<const qint32 * >( ptr + i * context.pointRecordSize() + context.zOffset() );
+  return context.offset().z() + context.scale().z() * iz;
+}
+
+struct MapIndexedPointCloudNode
+{
+  typedef QVector<QMap<QString, QVariant>> result_type;
+
+  MapIndexedPointCloudNode( QgsPointCloudRequest &request, QgsPointCloudRenderContext &context,
+                            QgsGeometry &extentGeometry, const QgsDoubleRange &zRange, QgsPointCloudIndex *index, int pointsLimit )
+    : mRequest( request ), mContext( context ), mExtentGeometry( extentGeometry ), mZRange( zRange ), mIndex( index ), mPointsLimit( pointsLimit )
+  {
+
+  }
+
+  QVector<QMap<QString, QVariant>> operator()( const IndexedPointCloudNode &n )
+  {
+    QVector<QMap<QString, QVariant>> acceptedPoints;
+    std::unique_ptr<QgsPointCloudBlock> block( mIndex->nodeData( n, mRequest ) );
+
+    if ( !block || pointsCount == mPointsLimit )
+      return acceptedPoints;
+
+    const char *ptr = block->data();
+    QgsPointCloudAttributeCollection blockAttributes = block->attributes();
+    const std::size_t recordSize = blockAttributes.pointRecordSize();
+    mContext.setAttributes( block->attributes() );
+    for ( int i = 0; i < block->pointCount(); ++i )
+    {
+      double x, y, z;
+      _pointXY( mContext, ptr, i, x, y );
+      z = _pointZ( mContext, ptr, i );
+      QgsPointXY pointXY( x, y );
+
+      if ( pointsCount < mPointsLimit && mExtentGeometry.contains( &pointXY ) && mZRange.contains( z ) )
```

Comment:
```suggestion
      QgsPoint pointXY( x, y );

      if ( pointsCount < mPointsLimit && extentEngine->intersects( &pointXY ) && mZRange.contains( z ) )
```

Reply:
(Using prepared geometries for all the containments checks will be MUCH faster)

Reply:
Wow that's much faster than my initial implementation especially for circle points. Thanks